### PR TITLE
Allow private or public claims to be parsed as Json objects

### PR DESCRIPTION
**Pull request by evbo**
A common use-case of Jwt-Scala (and JWT at large) is developing so-called "public" or "private" claims and or header params.  These are functionally no different than "reserved" claims/header params, other than that they have no standard definition and may only be distinguished within your network or niche industry. "issuer", "subject", "audience" etc. are all examples of reserved claims, whereas "username" is a fairly common example of a non-reserved claim. 

Given that there may be many of these public/private claims, it would save a lot of boilerplate code not having to parse them separate from how reserved claims are parsed. In other words, it makes sense to either extend or compose with reserved jwt claims/header params, so that you can customize for your public or private network's needs. 

Here is an example where reserved headers along with a private username claim is used:

```scala
// define your network-specific claims, and compose them with the usual reservedClaims
case class JwtPrivateClaim(user: Option[String] = None, reservedClaims: JwtClaim = JwtClaim()) {

// merge your json definition along with the reserved claims too
  def toJson: String = JwtUtils.mergeJson(JwtUtils.hashToJson(Seq(
      "user" -> user,
    ).collect {
    case (key, Some(value)) => (key -> value)
    }), reservedClaims.toJson)
}

// create a parser with claim type set to the one you just defined
case object JwtJson4sPrivate extends JwtJson4sParser[JwtHeader, JwtPrivateClaim] {
  override protected def parseClaim(claim: String): JwtPrivateClaim = {
    val claimJson = parse(claim)
    val jwtReservedClaim: JwtClaim = readClaim(claimJson)
    val content = parse(jwtReservedClaim.content)
    JwtPrivateClaim(extractString(content, "user"), jwtReservedClaim.copy(content = "{}"))
  }

// here is the only boilerplate (if you don't have private header params too):
  override protected def parseHeader(header: String): JwtHeader = readHeader(parse(header))

// marginal boilerplate as well I suppose
  override protected def extractExpiration(claim: JwtPrivateClaim): Option[Long] = claim.reservedClaims.expiration
  override protected def extractNotBefore(claim: JwtPrivateClaim): Option[Long] = claim.reservedClaims.notBefore
}

// here is how you marshal your claims:
trait JsonSupport extends SprayJsonSupport with DefaultJsonProtocol {
  implicit val jwtReservedClaimFormat = jsonFormat8(JwtClaim)
  implicit val jwtPrivateClaimFormat = jsonFormat2(JwtPrivateClaim)
}
```

you can then use the decodeAll method as before, now with your fully objectified claims:

```scala
val decoded: Try[(JwtHeader, JwtPrivateClaim, String)] = JwtJson4sPrivate.decodeAll(token)
```

This PR only changes modifiers and type generics, no functional code change was made and unit tests all pass still.

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -8,7 +8,7 @@ import scala.sys.process._
 import com.typesafe.sbt.pgp.PgpKeys._
 
 val previousVersion = "0.15.0"
-val buildVersion = "0.16.0"
+val buildVersion = "0.16.1"
```
Comment (pauldijou) [inline build.sbt]: If merged, this will be `0.17.0`, you also need to update `previousVersion`.

File: json/circe/src/main/scala/JwtCirce.scala [hunk 2]
```diff
@@ -11,49 +11,48 @@ import pdi.jwt.exceptions.JwtNonStringException
...
     }
   }
+}
 
+case object JwtCirce extends JwtCirceParser[JwtHeader, JwtClaim] {
   protected def parseHeader(header: String): JwtHeader = {
     val cursor = parse(header).hcursor
     JwtHeader(
-        algorithm = getAlg(cursor)
+      algorithm = getAlg(cursor)
```
Comment (pauldijou) [inline json/circe/src/main/scala/JwtCirce.scala]: Indentation

File: json/circe/src/main/scala/JwtCirce.scala [hunk 3]
```diff
@@ -11,49 +11,48 @@ import pdi.jwt.exceptions.JwtNonStringException
...
+  protected def parseClaim(claim: String): JwtClaim = {
+    val cursor = parse(claim).hcursor
+    val contentCursor = List("iss", "sub", "aud", "exp", "nbf", "iat", "jti").foldLeft(cursor) { (cursor, field) =>
+      cursor.downField(field).delete.success match {
+        case Some(newCursor) => newCursor
+        case None => cursor
+      }
+    }
+    JwtClaim(
+      content = contentCursor.top.asJson.noSpaces
```
Comment (pauldijou) [inline json/circe/src/main/scala/JwtCirce.scala]: Indentation

File: json/json4s-native/src/main/scala/JwtJson4sNative.scala [hunk 4]
```diff
@@ -21,7 +21,9 @@ object JwtJson4s extends JwtJson4sCommon with JwtJson4sImplicits {
   protected def stringify(value: JObject): String = compact(render(value))
 
   protected implicit val formats = Serialization.formats(NoTypeHints)
+}
 
+case object JwtJson4s extends JwtJson4sParser[JwtHeader, JwtClaim] {
   protected def parseHeader(header: String): JwtHeader = readHeader(parse(header))
   protected def parseClaim(claim: String): JwtClaim = readClaim(parse(claim))
-}
+}
```
Comment (pauldijou) [inline json/json4s-native/src/main/scala/JwtJson4sNative.scala]: New line

File: json/upickle/src/main/scala/JwtUpickle.scala [hunk 5]
```diff
@@ -21,8 +21,9 @@ object JwtUpickle extends JwtJsonCommon[Js.Value] with JwtUpickleImplicits {
...
     case _ => None
   }
+}
 
+case object JwtUpickle extends JwtUpickleParser[JwtHeader, JwtClaim] {
   protected def parseHeader(header: String): JwtHeader = read[JwtHeader](header)
   protected def parseClaim(claim: String): JwtClaim = read[JwtClaim](claim)
-
-}
+}
```
Comment (pauldijou) [inline json/upickle/src/main/scala/JwtUpickle.scala]: New line

**Review by pauldijou — CHANGES_REQUESTED** [review]
~~Did you manage to run the tests in the end?~~ You did 👍 

Why are you using `case object` rather than just `object`?

Otherwise, seems pretty good.

**Comment by evbo** [comment]
The benefits to using case object instead of object are very slim, if hardly there at all. I can just use object, but in case you're curious here's a breakdown of the differences: https://stackoverflow.com/a/5271730/1080804

I'll make the other corrections as well. Thanks!

**Review by pauldijou — APPROVED** [review]

**Comment by pauldijou** [comment]
Do you think you could write a bit of doc about it? Like how it works basically and an example on how to create your own custom header/claim. Best way to do that would be to write a `tut` markdown file. You can put it [with all others](https://github.com/pauldijou/jwt-scala/tree/master/docs/src/main/tut/_includes/tut). Also, add it [as a sample](https://github.com/pauldijou/jwt-scala/tree/master/docs/src/site/samples). Run `docs/tut` to test if it compiles, don't bother testing the website, I will add a link to it after merging.

**Comment by evbo** [comment]
I committed the docs yesterday - sorry forgot to leave a comment about it. Let me know if there's anything else you need to merge.

**Comment by evbo** [comment]
@pauldijou are we good to merge?

**Comment by pauldijou** [comment]
I did merge it but wanted to understand better. Is the goal just to add custom fields to the `JwtClaim`? Because you could already do that **before** this PR right? Check the [play-json](http://pauldijou.fr/jwt-scala/samples/jwt-play-json/) sample, it shows how to add a `user` custom field.

So, I don't really want the markdown documentation to focus on that. You should just use the normal claim for such purpose. Is there any other benefit for this code?

**Comment by evbo** [comment]
Not only are you 100% right, but the code is actually simpler if you just parse the `content` param rather than having to implement all the boilerplate my PR was proposing.

Whoops, complete oversight on my part. Sorry, but I think this merge should be reverted.

Here's a much cleaner approach to take now that I better understand:

```
import org.json4s._
import org.json42.jackson.JsonMethods.parse

case class JwtPrivateClaim(user: Option[String] = None)

val privateClaims = parse(JwtJson4s.decodeAll(token, JwtOptions(signature = false))._2.content)
    .extract[JwtPrivateClaim]

authorization(privateClaims) ...
```

For what it's worth, the only novetly my PR introduced was allowing custom header fields, because unlike `JwtClaim`, `JwtHeader` does not have a `content` property. Maybe it should to allow similar customization as there are several other header params not implemented that someone else may want to have (for instance I needed "kid" but there's others - ["jwk", "x5u", etc](https://tools.ietf.org/html/rfc7515#section-4.1)).

**Pull request merged**